### PR TITLE
Support model streaming responses with new `Stream` gRPC service

### PR DESCRIPTION
**Pull request by spillai**
This PR adds a new `Stream` gRPC service that allows for streaming responses from the server. This is useful for models that produce a stream of outputs such as text, video frames or audio chunks. The `Stream` service is implemented as a uni-directional output stream with a streaming response.

New models:
 - `llama2_chat` - a simple chatbot model that uses the `transformers` `AutoModelForCausalLM` with a `chat` submethod. New models include: `meta-llama/Llama-2-7b-chat-hf`, `HuggingFaceH4/zephyr-7b-beta`, `HuggingFaceH4/tiny-random-LlamaForCausalLM` (for testing purposes).

Other changes:
 - refactored names for all exceptions without `Nos` prefix

## Checks

- [ ] `make lint`: I've run `make lint` to lint the changes in this PR.
- [ ] `make test`: I've made sure the tests (`make test-cpu` or `make test`) are passing.
- Additional tests:
   - [ ] Benchmark tests (when contributing new models)
   - [ ] GPU/HW tests

### CODE REVIEW

File: nos/client/grpc.py [hunk 1]
```diff
@@ -357,6 +358,18 @@ def Run(
...
         return module(**inputs, _method=method)
 
+    def Stream(
+        self,
+        model_id: str,
+        inputs: Dict[str, Any],
+        method: str = None,
+        shm: bool = False,
+    ) -> Iterable[nos_service_pb2.GenericResponse]:
+        """Run module in streaming mode."""
```
Comment (outtanames) [inline nos/client/grpc.py:368]: [nit] could use a longer docstring

File: nos/client/grpc.py [hunk 2]
```diff
@@ -612,40 +626,59 @@ def __call__(self, _method: str = None, **inputs: Dict[str, Any]) -> Dict[str, A
...
         except Exception as e:
             logger.error(f"Failed to decode response [model={self.id}, e={e}]")
-            raise NosClientException(f"Failed to decode response [model={self.id}, e={e}]", e)
+            raise ClientException(f"Failed to decode response [model={self.id}, e={e}]", e)
         if NOS_PROFILING_ENABLED:
             logger.debug(f"Decoded response [model={self.id}, elapsed={(time.perf_counter() - st) * 1e3:.1f}ms]")
         return response
+
+
+@dataclass
```
Comment (outtanames) [inline nos/client/grpc.py:674]: [nit] should this live in common maybe?

File: nos/models/llama2_chat.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        temperature: float = 0.6,
+        top_p: float = 0.9,
+        top_k: int = 50,
+        repetition_penalty: float = 1.2,
+        num_beams: int = 1,
+    ) -> Iterable[str]:
+        """Chat with the model."""
+        conversation = []
+        if system_prompt:
+            conversation.append({"role": "system", "content": system_prompt})
```
Comment (outtanames) [inline nos/models/llama2_chat.py:87]: [nit] this is a bit hard to follow. Can we define a template with fields for conversation entries?

**Review by outtanames — APPROVED** [review]
:100:

**Comment by spillai** [comment]
@outtanames Will add these requested changes to the subsequent PR since there was a bit of refactor for the OpenAI SDK
